src/gui/guiEngine.cpp

```diff
@@ -213,6 +215,57 @@ GUIEngine::GUIEngine(JoystickController *joystick,
 	m_menu = NULL;
 }
 
+
+/******************************************************************************/
+std::string findLocaleFileInMods(const std::string &path, const std::string &filename)
+{
+	std::vector<ModSpec> mods = flattenMods(getModsInPath(path, "root", true));
+
+	for (const auto &mod : mods) {
+		std::string ret = mod.path + DIR_DELIM "locale" DIR_DELIM + filename;
+		if (fs::PathExists(ret)) {
+			return ret;
+		}
+	}
+
+	return "";
+}
+
+/******************************************************************************/
+Translations *GUIEngine::getContentTranslations(const std::string &path,
+		const std::string &domain, const std::string &lang_code)
+{
+	if (domain.empty() || lang_code.empty())
+		return nullptr;
+
+	std::string filename = domain + "." + lang_code + ".tr";
+	std::string key = path + DIR_DELIM "locale" DIR_DELIM + filename;
+
+	auto it = translations.find(key);
```

Comment:
This will result in all the translations being loaded into memory on menu start up, and kept there until the menu closes

My suggestion would be to only cache the last accessed mod's translation, as this is the simplest approach

Reply:
You could pass a new parameter to `loadTranslation` indicating the origin of the translation so that it could be unloaded automatically.

Reply:
What do you mean by "origin of the translation"? The modname or the fact this is the mainmenu? This is all unloaded when the user exits the menu (ie: by starting a game)

Reply:
Here's what it might look like to only store one translation at a time (warning: this would result in bad data if getContentTranslations is called with a different key whilst a point to last_translation is retained)

```cpp.patch
diff --git a/src/gui/guiEngine.cpp b/src/gui/guiEngine.cpp
index b5adde76b..2b2c09205 100644
--- a/src/gui/guiEngine.cpp
+++ b/src/gui/guiEngine.cpp
@@ -232,9 +232,8 @@ Translations *GUIEngine::getContentTranslations(const std::string &path,
 	std::string filename = domain + "." + lang_code + ".tr";
 	std::string key = path + DIR_DELIM "locale" DIR_DELIM + filename;
 
-	auto it = translations.find(key);
-	if (it != translations.end())
-		return &it->second; // Already loaded
+	if (key == last_translations_key)
+		return &last_translations;
 
 	std::string trans_path = key;
 	ContentType type = getContentType(path);
@@ -246,15 +245,15 @@ Translations *GUIEngine::getContentTranslations(const std::string &path,
 	if (trans_path.empty())
 		return nullptr;
 
-	// [] will create an entry
-	auto *ret = &translations[key];
+	last_translations_key = key;
+	last_translations = {};
 
 	std::string data;
 	if (fs::ReadFile(trans_path, data)) {
-		ret->loadTranslation(data);
+		last_translations.loadTranslation(data);
 	}
 
-	return ret;
+	return &last_translations;
 }
 
 /******************************************************************************/
diff --git a/src/gui/guiEngine.h b/src/gui/guiEngine.h
index 4264bb6ed..e4a84fbb8 100644
--- a/src/gui/guiEngine.h
+++ b/src/gui/guiEngine.h
@@ -176,7 +176,8 @@ class GUIEngine {
 			const std::string &domain, const std::string &lang_code);
 
 private:
-	std::unordered_map<std::string, Translations> translations;
+	std::string last_translations_key;
+	Translations last_translations;
 
 	/** find and run the main menu script */
 	bool loadMainMenuScript();
```

Reply:
Ok, I'm going to go with the simple bug-free thing and remove caching from getContentTranslations

Reply:
Ok, pushed a change - it only caches the last translation now